core/modules/replica/PurgeJob.h

```diff
@@ -0,0 +1,266 @@
+/*
+ * LSST Data Management System
+ * Copyright 2017 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_REPLICA_PURGEJOB_H
+#define LSST_QSERV_REPLICA_PURGEJOB_H
+
+/// PurgeJob.h declares:
+///
+/// struct PurgeJobResult
+/// class  PurgeJob
+///
+/// (see individual class documentation for more information)
+
+// System headers
+#include <atomic>
+#include <functional>
+#include <list>
+#include <map>
+#include <string>
+
+// Qserv headers
+#include "replica/DeleteReplicaJob.h"
+#include "replica/FindAllJob.h"
+#include "replica/ReplicaInfo.h"
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+/**
+ * The structure PurgeJobResult represents a combined result received
+ * from worker services upon a completion of the job.
+ */
+struct PurgeJobResult {
+
+    /// Results reported by workers upon the successfull completion
+    /// of the corresponidng jobs
+    std::list<ReplicaInfo> replicas;
+
+    /// Results groupped by: chunk number, database, worker
+    std::map<unsigned int,                  // chunk
+             std::map<std::string,          // database
+                      std::map<std::string, // worker
+                               ReplicaInfo>>> chunks;
```

Comment:
This is interesting. I've seen cases in Python where nested containers are used where really a data structure was more appropriate. It got way out of hand. This seems reasonable, but maybe only just 😆

Reply:
I myself hate nested maps :-) Unfortunately I still need some way (a data structure) to organize results of a job in a compact and convenient way to use. After I realized how horrible (and error prone) I came up with a design which I called: SemanticMaps.h
Just scroll down that header file (SemanticMaps.h). It shows how that ugly nested map is replaced with something which allows me to do:

ReplicaInfo& info = map.chunk(123).database("db-name").worker("host-name");

instead of:

ReplicaInfo& info = map[123]["db-name"]["host-name"];

The later is obviously very unsafe because one can do:

ReplicaInfo& info = map[123]["host-name"]["db-name"];

and be very surprised with results.

Reply:
A think is that I'm still in a process of migrating the remaining nested maps to that "SemanticMap".

And, no, no, no - I still don't like that Boost thing MultiIndex :-)